pkg/operators/combiner/combiner.go

```diff
@@ -0,0 +1,304 @@
+			}
+
+			targetCount++
+
+			// TODO: Support Append multiple packets at once?
+			for i := 0; i < inPacket.Len(); i++ {
+				combinedPacket.Append(inPacket.Get(i))
+			}
+
+			// For data sources that doesn't have an interval, we wait for data
+			// from all targets before emitting the combined data.
+			if interval == 0 && targetCount == o.targets {
+				if err := emitAndAllocate(); err != nil {
+					gadgetCtx.Logger().Errorf("emitting and allocating data: %s", err)
+				}
+				return
+			}
+		case <-c:
+			if interval == 0 {
+				gadgetCtx.Logger().Warnf("Data is incomplete: timeout waiting for data from all targets (%d/%d)",
+					targetCount, o.targets)
+			}
+
+			if err := emitAndAllocate(); err != nil {
+				gadgetCtx.Logger().Errorf("emitting and allocating data: %s", err)
+				return
+			}
+		}
+	}
+}
+
+func (o *combinerOperatorInstance) PreStart(gadgetCtx operators.GadgetContext) error {
+	o.done = make(chan struct{})
+
+	for _, ds := range gadgetCtx.GetDataSources() {
+		if ds.Type() != datasource.TypeArray {
+			continue
+		}
+
+		interval, err := getFetchAnnotation(ds)
+		if err != nil {
+			return fmt.Errorf("getting fetch annotation for ds %s: %w", ds.Name(), err)
+		}
+
+		// Disable original data source to avoid other operators subscribing to it
+		ds.SetRequested(false)
+
+		// Register a new data source that will emit the combined data
+		combinedDs, err := gadgetCtx.RegisterDataSource(
+			datasource.TypeArray,
+			fmt.Sprintf("%s-%s", DataSourcePrefix, ds.Name()),
+		)
+		if err != nil {
+			return fmt.Errorf("registering combiner data source for %s: %w", combinedDs.Name(), err)
+		}
+
+		gadgetCtx.Logger().Debugf("%s: registered ds %s", o.Name(), combinedDs.Name())
+
+		// Use the same fields and annotations as the original data source
+		ds.CopyFieldsTo(combinedDs)
+		for k, v := range ds.Annotations() {
+			combinedDs.AddAnnotation(k, v)
+		}
+
+		o.pktBuffers[ds.Name()] = make(chan datasource.PacketArray, o.targets)
+		go o.forwardData(gadgetCtx, ds.Name(), interval, combinedDs)
+
+		gadgetCtx.Logger().Debugf("%s: subscribing to %s", o.Name(), ds.Name())
+
+		ds.SubscribePacket(func(source datasource.DataSource, packet datasource.Packet) error {
+			// Use Marshal/Unmarshal to create a deep copy
+			b, _ := proto.Marshal(packet.Raw())
+			pArray, err := combinedDs.NewPacketArrayFromRaw(b)
+			if err != nil {
+				return fmt.Errorf("creating packet array from raw: %w", err)
+			}
+
+			// Send the packet to the combiner data source
+			o.pktBuffers[ds.Name()] <- pArray
+
+			return nil
+		}, Priority)
+	}
+
+	return nil
+}
+
+func (o *combinerOperatorInstance) Start(gadgetCtx operators.GadgetContext) error {
+	return nil
+}
+
+func (o *combinerOperatorInstance) Stop(gadgetCtx operators.GadgetContext) error {
+	if o.done != nil {
+		close(o.done)
+		o.done = nil
+	}
+	if o.pktBuffers != nil {
+		for _, buf := range o.pktBuffers {
+			if buf != nil {
+				close(buf)
+			}
+		}
+		o.pktBuffers = nil
+	}
+	return nil
+}
```

Comment:
I don't think it'll be called more than once, but I don't know the approach we want to have here. If for any reason we call it twice or we stop a instance that hasn't been started, it will crash. Do we want to protect operators against that or we prefer to rely that the framework is working properly?

Reply:
Good point, I think we need to clarify it. IMHO Stop() should never be called twice, I think that easy to do. I'm worried about the case when Stop() is called because Start() failed and some of the things in the instance could have not been initialized, in that case the test will make sense. 

Let's keep it as it's and make a pass on all operators later on.